### PR TITLE
core(minification-estimator): minify nested template literals in JavaScript

### PR DESCRIPTION
**Pull request by adamraine**
Fixes https://github.com/GoogleChrome/lighthouse/issues/11295

### CODE REVIEW

File: lighthouse-core/lib/minification-estimator.js [hunk 1]
```diff
@@ -50,6 +50,12 @@ function computeTokenLength(content, features) {
   let isInRegexCharacterClass = false;
   let stringOpenChar = null;
 
+  /**
+   * Acts as stack for brace tracking.  Entries are true when innermost brace is template literal.
+   * @type boolean[]
```
Comment (connorjclark) [inline lighthouse-core/lib/minification-estimator.js:55]:
```suggestion
   * @type {boolean[]}
```

(not a nit: without the braces, types-in-jsdoc doesn't work)

File: lighthouse-core/test/lib/minification-estimator-test.js [hunk 2]
```diff
@@ -227,5 +227,19 @@ describe('minification estimator', () => {
...
+      // Basic nested literals
+      const nestedTemplates = 'window.myString=`foo${` bar ${` baz ${` bam `} `} `} `';
+      expect(computeJSTokenLength(nestedTemplates)).toEqual(nestedTemplates.length);
+
+      // Can get rid of 5 spaces after normal code braces
+      const nestedWithCode = 'window.myString=`foo${` bar ${{}     }`}`';
+      expect(computeJSTokenLength(nestedWithCode)).toEqual(nestedWithCode.length - 5);
+
+      // Ignore braces in string
+      const nestedTemplatesBrace = 'window.myString=`{foo${` }bar ${` baz ${` bam `} `} `} `';
```
Comment (patrickhulce) [inline lighthouse-core/test/lib/minification-estimator-test.js:241]: nice cases! maybe just one more that flips back and forth between brace types?

```js
window.myString=`${({foo: bar.map(() => ({baz: `${'}'}`}))})}`

**Review by patrickhulce — CHANGES_REQUESTED** [review]
thanks @adamraine nice diving in here!

File: lighthouse-core/lib/minification-estimator.js [hunk 3]
```diff
@@ -129,6 +141,18 @@ function computeTokenLength(content, features) {
         isInRegex = true;
         // Regex characters count
         totalTokenLength++;
+      } else if (char === '{' && templateLiteralDepth) {
```
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/lib/minification-estimator.js:144]: Correct, this is just so we are not managing brace depth in normal javascript code outside template literals.

File: lighthouse-core/test/lib/minification-estimator-test.js (hunk 2 above)
Comment (adamraine) [reply lighthouse-core/test/lib/minification-estimator-test.js:241]: Is this test code fully minified?

Comment (patrickhulce) [reply lighthouse-core/test/lib/minification-estimator-test.js:241]: no there should be 4 spaces that can be removed

Comment (adamraine) [reply lighthouse-core/test/lib/minification-estimator-test.js:241]: I removed the spaces for this case, would it be better to keep them an adjust the expectation?

**Review by patrickhulce — APPROVED** [review]

File: lighthouse-core/test/lib/minification-estimator-test.js (hunk 2 above)
Comment (patrickhulce) [reply lighthouse-core/test/lib/minification-estimator-test.js:241]: yeah I would probably prefer that so it recognizes it's in JS

Comment (patrickhulce) [reply lighthouse-core/test/lib/minification-estimator-test.js:241]: Not a big deal though, so I'll leave this up to you @adamraine if you want to address before merge

**Pull request merged**